### PR TITLE
Fix pagination on bulk delete

### LINKED ISSUE #4794: warning when deleting last element on last page in data table
**What you were expecting:**
no warning message in the console

**What happened instead:**
there is a warning message in the console: `The page prop of a TablePagination is out of range`

**Steps to reproduce:**
delete the last element on the last page of the data table

https://youtu.be/2pM-VHZn5BY

**Environment**

* React-admin version: 3.5.0
* React version:16.13.1
* Browser: chrome
* Stack trace (in case of a JS error):

```
Warning: Failed prop type: Material-UI: The page prop of a TablePagination is out of range (0 to 117, but page is 118).
    in ForwardRef(TablePagination) (created by WithStyles(ForwardRef(TablePagination)))
    in WithStyles(ForwardRef(TablePagination)) (created by Pagination)
    in Pagination
    in Pagination (at accounts.js:33)
    in div (created by ForwardRef(Paper))
    in ForwardRef(Paper) (created by WithStyles(ForwardRef(Paper)))
    in WithStyles(ForwardRef(Paper)) (created by ForwardRef(Card))
    in ForwardRef(Card) (created by WithStyles(ForwardRef(Card)))
    in WithStyles(ForwardRef(Card)) (created by ListView)
    in div (created by ListView)
    in div (created by ListView)
    in ListView (created by List)
    in List (at accounts.js:33)
    in AccountList (created by WithPermissions)
    in WithPermissions (created by Context.Consumer)
    in Route (created by ResourceRoutes)
    in Switch (created by ResourceRoutes)
    in ResourceRoutes (created by Resource)
    in Resource (at App.js:12)
    in Route (created by RoutesWithLayout)
    in Switch (created by RoutesWithLayout)
    in RoutesWithLayout (created by Context.Consumer)
    in div (created by Layout)
    in main (created by Layout)
    in div (created by Layout)
    in div (created by Layout)
    in Layout (created by WithStyles(Layout))
    in WithStyles(Layout) (created by Context.Consumer)
    in withRouter(WithStyles(Layout)) (created by ConnectFunction)
    in ConnectFunction (created by LayoutWithTheme)
    in ThemeProvider (created by LayoutWithTheme)
    in LayoutWithTheme (created by Context.Consumer)
    in Route (created by CoreAdminRouter)
    in Switch (created by CoreAdminRouter)
    in div (created by CoreAdminRouter)
    in CoreAdminRouter (created by Context.Consumer)
    in Route (created by CoreAdminUI)
    in Switch (created by CoreAdminUI)
    in CoreAdminUI (created by AdminUI)
    in AdminUI (created by Admin)
    in Router (created by ConnectedRouter)
    in ConnectedRouter (created by Context.Consumer)
    in ConnectedRouterWithContext (created by ConnectFunction)
    in ConnectFunction (created by CoreAdminContext)
    in TranslationProvider (created by CoreAdminContext)
    in Provider (created by CoreAdminContext)
    in CoreAdminContext (created by AdminContext)
    in AdminContext (created by Admin)
    in Admin (at App.js:11)
    in App (at src/index.js:9)
    in StrictMode (at src/index.js:8)
console.<computed> @ index.js:1
r @ react_devtools_backend.js:6
printWarning @ checkPropTypes.js:20
checkPropTypes @ checkPropTypes.js:82
validatePropTypes @ react.development.js:1714
createElementWithValidation @ react.development.js:1813
WithStyles @ withStyles.js:79
renderWithHooks @ react-dom.development.js:14803
updateForwardRef @ react-dom.development.js:16816
beginWork @ react-dom.development.js:18645
beginWork$1 @ react-dom.development.js:23179
performUnitOfWork @ react-dom.development.js:22154
workLoopSync @ react-dom.development.js:22130
performSyncWorkOnRoot @ react-dom.development.js:21756
(anonymous) @ react-dom.development.js:11089
unstable_runWithPriority @ scheduler.development.js:653
runWithPriority$1 @ react-dom.development.js:11039
flushSyncCallbackQueueImpl @ react-dom.development.js:11084
flushSyncCallbackQueue @ react-dom.development.js:11072
discreteUpdates$1 @ react-dom.development.js:21893
discreteUpdates @ react-dom.development.js:806
dispatchDiscreteEvent @ react-dom.development.js:4168
```

### PR DESCRIPTION
**Pull request by JulienMattiussi**
Closes #4794 

## Issue 1 : 

Fix MUI warning when deleting last items of last page by skipping the rendering before fallback to the previous page.

image

## Issue 2 :

Avoid List to always fallback on Page 1 when bulk-deleting some items.

Fallback1

### CODE REVIEW

File: packages/ra-core/src/controller/useListController.spec.tsx [hunk 1]
```diff
@@ -148,7 +148,7 @@ describe('useListController', () => {
             ).toEqual({ q: true });
         });
 
-        it('should update data if permanent filters change', () => {
+        it.only('should update data if permanent filters change', () => {
```
Comment (fzaninotto) [inline packages/ra-core/src/controller/useListController.spec.tsx:151]: .only detected 😱

File: packages/ra-core/src/dataProvider/useGetList.ts [hunk 2]
```diff
@@ -71,11 +71,18 @@ const useGetList = <RecordType = Record>(
         { type: 'getList', resource, payload: { pagination, sort, filter } },
         options,
         // data selector (may return [])
+        // We don't want the list to reset to first page after a bulk delete if the current page still exists
+        // But CrudDeleteMany reset all "cachedRequests" for this resource.
+        // It caused ids to be empty, and useListController to automatically reset to page 1
+        // Now, it "cachedRequests" is undefined, we try default ids before setting an empty ids array
```
Comment (fzaninotto) [inline packages/ra-core/src/dataProvider/useGetList.ts:77]: we already have such a system in `useListController`. I think that's where it belongs, not in `useGetList`.

File: packages/ra-ui-materialui/src/list/Pagination.js [hunk 3]
```diff
@@ -25,16 +25,21 @@ const Pagination = ({
...
+    ]);
+
     useEffect(() => {
         if (page < 1 || isNaN(page)) {
             setPage(1);
+        } else if (page > getNbPages()) {
+            setPage(getNbPages());
         }
-    }, [page, setPage]);
+    }, [page, setPage, getNbPages, total, perPage]);
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/list/Pagination.js:39]: no need for `perPage` or `total` as it's already in the `useCallback` dependency array above.

File: packages/ra-core/src/dataProvider/useGetList.ts (hunk 2 above)
Comment (JulienMattiussi) [reply packages/ra-core/src/dataProvider/useGetList.ts]: Fixed, and then removed because already fixed in master since I opened this PR ...

**Comment by Luwangel** [comment]
Hi! You have a merge conflict. Could you rebase it?

File: packages/ra-core/src/controller/field/useReferenceArrayFieldController.ts [hunk 4]
```diff
@@ -194,6 +194,8 @@ const useReferenceArrayFieldController = ({
         sort.order,
     ]);
 
+    const total = finalIds.length;
```
Comment (JulienMattiussi) [inline packages/ra-core/src/controller/field/useReferenceArrayFieldController.ts:197]: no more usefull now IMO

File: packages/ra-core/src/controller/field/useReferenceManyFieldController.ts [hunk 5]
```diff
@@ -1,5 +1,5 @@
 import get from 'lodash/get';
-import { useCallback, useEffect, useRef } from 'react';
+import { useCallback, useMemo, useEffect, useRef } from 'react';
```
Comment (JulienMattiussi) [inline packages/ra-core/src/controller/field/useReferenceManyFieldController.ts:2]: no more used

File: packages/ra-core/src/controller/useListController.ts [hunk 6]
```diff
@@ -184,15 +184,31 @@ const useListController = <RecordType = Record>(
...
             queryModifiers.setPage(1);
+        } else if (!loading && query.page > totalPages) {
+            // Query for a page out of bounds, set page to the last existing page
+            // It occurs when deleting the last element of the last page
+            queryModifiers.setPage(totalPages);
         }
-    }, [loading, query.page, finalIds, queryModifiers, total, defaultIds]);
+    }, [
+        loading,
+        query,
```
Comment (JulienMattiussi) [inline packages/ra-core/src/controller/useListController.ts:205]: why changing query.page to query here ? only query.page is used in the useEffect

File: packages/ra-core/src/controller/useListParams.spec.ts [hunk 7]
```diff
@@ -156,4 +156,30 @@ describe('useListParams', () => {
...
+
+            expect(result).toEqual(2);
+        });
+
+        it('should not return the default number when passing "0"', () => {
+            const result = getNumberOrDefault('0', 2);
+
+            expect(result).toEqual(0);
+        });
+    });
```
Comment (JulienMattiussi) [inline packages/ra-core/src/controller/useListParams.spec.ts:184]: you should add the test that passing 0 return 0

File: packages/ra-core/src/controller/useListParams.ts [hunk 8]
```diff
@@ -358,9 +359,13 @@ export const getQuery = ({
...
-        ? parseInt(possibleNumber, 10)
-        : possibleNumber) || defaultValue;
+) => {
+    const parsedNumber =
+        typeof possibleNumber === 'string'
+            ? parseInt(possibleNumber, 10)
+            : possibleNumber;
+
+    return isNaN(parsedNumber) ? defaultValue : parsedNumber;
+};
```
Comment (JulienMattiussi) [inline packages/ra-core/src/controller/useListParams.ts:369]: :+1:

File: packages/ra-ui-materialui/src/list/Pagination.spec.js [hunk 9]
```diff
@@ -43,19 +43,50 @@ describe('<Pagination />', () => {
             expect(queryByText('ra.navigation.no_results')).toBeNull();
         });
 
-        it('should not display a pagination limit on an out of bounds page', () => {
+        it('should display a pagination limit on an out of bounds page (more than total pages)', async () => {
```
Comment (JulienMattiussi) [inline packages/ra-ui-materialui/src/list/Pagination.spec.js:46]: why changing this title ?
the previous was true, we not display the pagination limit component, we display the last page

File: packages/ra-ui-materialui/src/list/Pagination.spec.js [hunk 10]
```diff
@@ -43,19 +43,50 @@ describe('<Pagination />', () => {
...
                     >
                         <Pagination />
                     </ListContext.Provider>
                 </ThemeProvider>
             );
-            // mui TablePagination displays a warning in that case, and that's normal
-            expect(queryByText('ra.navigation.no_results')).toBeNull();
+            // mui TablePagination displays no more a warning in that case
+            // Then useEffect fallbacks on a valid page
+            expect(queryByText('ra.navigation.no_results')).not.toBeNull();
```
Comment (JulienMattiussi) [inline packages/ra-ui-materialui/src/list/Pagination.spec.js:66]: don't understand why we display the "no_result" if we fallback on a valid page 1

File: packages/ra-ui-materialui/src/list/Pagination.tsx [hunk 11]
```diff
@@ -77,7 +75,8 @@ const Pagination: FC<PaginationProps> = props => {
         [translate]
     );
 
-    if (total === 0) {
+    // Avoid rendering TablePagination if "page" value is invalid
+    if (total === 0 || page < 0 || page > totalPages) {
```
Comment (JulienMattiussi) [inline packages/ra-ui-materialui/src/list/Pagination.tsx:79]: Here, I think the test is `page < 1`, not `page < 0`
(The 0 based page number is in the handle function only)

File: packages/ra-core/src/controller/useListParams.spec.ts (hunk 7 above)
Comment (JulienMattiussi) [reply packages/ra-core/src/controller/useListParams.spec.ts:184]: my bad, your're very good

**Pull request merged**